app/src/main/java/dev/arkbuilders/arkmemo/repo/NotesRepoHelper.kt

```diff
@@ -0,0 +1,83 @@
+package dev.arkbuilders.arkmemo.repo
+
+import android.util.Log
+import dev.arkbuilders.arklib.ResourceId
+import dev.arkbuilders.arklib.data.index.Resource
+import dev.arkbuilders.arklib.data.index.RootIndex
+import dev.arkbuilders.arklib.user.properties.Properties
+import dev.arkbuilders.arklib.user.properties.PropertiesStorage
+import dev.arkbuilders.arklib.user.properties.PropertiesStorageRepo
+import dev.arkbuilders.arkmemo.models.Note
+import dev.arkbuilders.arkmemo.preferences.MemoPreferences
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import java.nio.file.Path
+import javax.inject.Inject
+import kotlin.io.path.deleteIfExists
+import kotlin.io.path.extension
+import kotlin.io.path.getLastModifiedTime
+import kotlin.io.path.moveTo
+import kotlin.io.path.name
+
+class NotesRepoHelper @Inject constructor(
+    private val memoPreferences: MemoPreferences,
+    private val propertiesStorageRepo: PropertiesStorageRepo
+) {
+
+    private lateinit var root: Path
+    private lateinit var propertiesStorage: PropertiesStorage
+
+    suspend fun init() {
+        root = memoPreferences.getNotesStorage()
+        propertiesStorage = propertiesStorageRepo.provide(RootIndex.provide(root))
+    }
+
+    suspend fun persistNoteProperties(resourceId: ResourceId, noteTitle: String) {
+        with(propertiesStorage) {
+            val properties = Properties(setOf(noteTitle), setOf())
+            setProperties(resourceId, properties)
+            persist()
+        }
+    }
+
+    fun renameResource(
+        note: Note,
+        tempPath: Path,
+        resourcePath: Path,
+        resourceId: ResourceId,
+    ) {
+        tempPath.moveTo(resourcePath)
+        note.resource = Resource(
+            id = resourceId,
+            name = resourcePath.fileName.name,
+            extension = resourcePath.extension,
+            modified = resourcePath.getLastModifiedTime()
+        )
+        Log.d("notes-repo", "resource renamed to ${resourcePath.name} successfully")
+    }
+
+    fun readProperties(id: ResourceId): UserNoteProperties {
+        val title = propertiesStorage.getProperties(id).titles.let {
+            if (it.isNotEmpty()) it.elementAt(0) else throw NoteTitlesException()
```

Comment:
@ShubertMunthali This is not backward-compatible with previous app version.
That means, when there're existing notes which do not have any title, there will be a crash.
Thrown exception should be handled more elegantly in this case.

![Screenshot 2023-12-05 at 07 43 21](https://github.com/ARK-Builders/ARK-Memo/assets/4281969/9bbbfa30-96a3-4371-b401-8deae1fdc48f)

Reply:
Should we not throw exception? Then we can get the title from content, that's what old Memo did

Reply:
> Should we not throw exception? Then we can get the title from content, that's what old Memo did

@ShubertMunthali If we throw it here, we should catch and handle it from the caller side, to avoid crash.
It's your call to decide it should be thrown out here or not, but IMHO there's no need to throw an exception here.